pysisyphus/calculators/CFOUR.py

```diff
@@ -0,0 +1,249 @@
+    mobj = re.search(regex, text, re.DOTALL)
+    energy = float(mobj.groups()[0])
+
+    return {"energy": energy}
+
+
+def parse_cfour_gradient(grd_fn, out_fn, coords_pysis_frame_3d):
+    results = {}
+    gradient = np.loadtxt(grd_fn, skiprows=1)
+    natoms = int(gradient.shape[0] / 2)
+    gradient = gradient[natoms:, 1:]
+
+    coords_comp_frame_3d = read_cfour_geom(out_fn)
+    gradient_rotated = rotate_gradient(
+        gradient, coords_pysis_frame_3d, coords_comp_frame_3d
+    ).flatten()
+
+    energy = parse_cfour_energy(out_fn)["energy"]
+    results["energy"] = energy
+    results["forces"] = -gradient_rotated
+
+    return results
+
+
+def read_cfour_geom(out_fn):
+    with open(out_fn) as handle:
+        text = handle.read()
+
+    regex = r"Coordinates used in calculation \(QCOMP\)(.+)Interatomic distance matrix"
+    floats = [cfour_float_regex] * 3
+    # Element symbol, atomic number, then x, y, z in bohr
+    line_regex = r"\s*[A-Z]+\s*\d+\s*" + r"\s*".join(floats)
+
+    mobj = re.search(regex, text, re.DOTALL)
+    geom = list()
+    for line in mobj.groups()[0].split("\n"):
+        mobj = re.match(line_regex, line.strip())
+        if not mobj:
+            continue
+        geom.append(mobj.groups())
+    geom_3d = np.array(geom, dtype=float)
+
+    return geom_3d
+
+
+def rotate_gradient(gradient, coords_pysis_frame_3d, coords_comp_frame_3d):
+    # Following http://nghiaho.com/?page_id=671
+    # Permalink: https://web.archive.org/web/20230128004529/http://nghiaho.com/?page_id=671
+
+    comp_frame_centroid = calc_centroid(coords_comp_frame_3d)
+    pysis_frame_centroid = calc_centroid(coords_pysis_frame_3d)
+
+    rot_matrix = calc_rot_matrix(
+        coords_comp_frame_3d,
+        coords_pysis_frame_3d,
+        comp_frame_centroid,
+        pysis_frame_centroid,
+    )
+
+    return gradient @ rot_matrix.T
+
+
+def calc_centroid(coords_3d):
+    centroid = np.sum(coords_3d, axis=0) / coords_3d.shape[0]
+    return centroid
+
+
+def calc_rot_matrix(
+    source_coords_3d, target_coords_3d, source_centroid, target_centroid
+):
+    H = (source_coords_3d - source_centroid).T @ (target_coords_3d - target_centroid)
+    assert H.shape == (3, 3)
+    U, S, V = np.linalg.svd(H)
+    R = V @ U.T
+
+    # Cover corner case
+    if np.linalg.det(R) < 0:
+        U, S, V = np.linalg.svd(R)
+        V[:, 2] = V[:, 2] * -1
+        R = V @ U.T
+
+    return R
+
+
+class CFOUR(Calculator):
+    conf_key = "cfour"
+
+    def __init__(
+        self,
+        cfour_input,
+        wavefunction_dump=True,
+        initden_file=None,
+        **kwargs,
+    ):
+        """CFOUR calculator.
+
+        Wrapper handling CFOUR ground state energy and gradient calculations.
+
+        Parameters
+        ----------
+        cfour_input : dict
+            CFOUR keywords and values. Note: "on" must be encapsulated in quotes to avoid being translated to True by YAML.
```

Comment:
Did you run `black` on this file? At least this line seems quite long; or does `black` not format long comments?!
Please add a linebreak at a sensible position.

Reply:
I did run black, but apparently it doesn't format long comments. I'm as surprised as you are. Fixing.